lib/AST/DiagnosticEngine.cpp

```diff
@@ -173,10 +173,12 @@ InFlightDiagnostic &InFlightDiagnostic::fixItRemove(SourceRange R) {
   // If we're removing something (e.g. a keyword), do a bit of extra work to
   // make sure that we leave the code in a good place, without extraneous white
   // space around its hole.  Specifically, check to see there is whitespace
-  // before and after the end of range.  If so, nuke the space afterward to keep
-  // things consistent.
+  // or a paren before and whitespace after the end of range.  If so, nuke
+  // the space afterward to keep things consistent.
+  char charBefore = extractCharBefore(SM, charRange.getStart());
+
   if (extractCharAfter(SM, charRange.getEnd()) == ' ' &&
-      isspace(extractCharBefore(SM, charRange.getStart()))) {
+      (isspace(charBefore) || charBefore == '(')) {
```

Comment:
I think this should be done at the diagnostic site - https://github.com/apple/swift/blob/master/lib/Parse/ParsePattern.cpp#L239

@CodaFi I think it might make sense to give the diagnostic here, maybe a bit re-worded, and suggest to either remove the keyword via note or add backticks as you suggested in JIRA since it's not really obvious that user really meant to use `var` or `let` as a label. WDYT?

Reply:
Actually, come to think of it, why don't we just allow this without a diagnostic?  `var` and `let` no longer mean anything in this position, and we already allow other keywords in this position:

```
func foo<T>(class x: T) -> T { 
  return x  
} 
```

Reply:
I suspect that allowing this could cause a lot of confusion. Keywords like `class` really couldn't have another reasonable meaning in this context, so I'm guessing that's why they're allowed without backticks. Allowing the use of `var` here without backticks would probably make people think the parameter is a var binding locally when it's still a let binding. That said, @CodaFi  that is a good point, if we don't care about the possibility of confusion there, it might make sense to just allow those names. Thoughts @xedin?

Reply:
I'm guessing that let/var where explicitly prohibited because of the meaning they have, maybe if we were to all them it would make sense to have a warning which says that keyword `let`/`var` in that position is interpreted as a label?

Reply:
That seems reasonable.  It's pretty much always a mistake when this shows up.  You can suppress that diagnostic with the backpacks.

Reply:
Yeah, add a warning and note with a fix to add backticks is what I'm thinking.

Reply:
Sounds good 👍

Reply:
Update on this, because it makes sense now to give a warning when people add unnecessary backpacks at the call site for let / var, this can actually stay how it was.